crates/wasmtime/src/config.rs

```diff
@@ -14,6 +17,105 @@ use wasmtime_environ::settings::{self, Configurable, SetError};
 use wasmtime_environ::{isa, isa::TargetIsa, Tunables};
 use wasmtime_jit::{native, CompilationStrategy, Compiler};
 use wasmtime_profiling::{JitDumpAgent, NullProfilerAgent, ProfilingAgent, VTuneAgent};
+use wasmtime_runtime::{ExportFunction, InstanceHandle, VMTrampoline};
+
+// This type needs to be public for `IntoFunc`, but is considered an internal implementation detail.
+#[doc(hidden)]
+pub struct HostFunc {
+    ty: FuncType,
+    instance: InstanceHandle,
+    trampoline: VMTrampoline,
+}
+
+impl HostFunc {
+    pub(crate) fn new(ty: FuncType, instance: InstanceHandle, trampoline: VMTrampoline) -> Self {
+        Self {
+            ty,
+            instance,
+            trampoline,
+        }
+    }
+
+    pub(crate) fn to_func(&self, store: &Store) -> Func {
+        let instance = StoreInstanceHandle {
+            store: store.clone(),
+            // This clone of the instance handle should be safe because it should not be deallocated
+            // until all configs that reference it are dropped.
+            // A config will not drop until all stores referencing the config are dropped.
```

Comment:
Does this mean that we need to check that this store is using an engine that is using this `HostFunc`'s config? is it possible for these things to get out of sync (which seems like it would lead to unsafety)

Reply:
I will add a debug assertion just to double check that the `HostFunc` is indeed related to the associated config.

Reply:
Actually, the lookup can't be done quickly as the `HostFunc` doesn't store the requisite information to query the passed store.

I think making this unsafe as suggested in the other comment is the best approach.